### ISSUE TITLE
Image Garbage Collection [GC] Non-Exhaustive

### ISSUE THREAD

**Issue by osterman** [kubernetes/kubernetes#29325]
## what
Image GC works, but very slowly, deleting only a handful of images every 5 minutes.

## observations

*  GC queries API for a subset of images. This subset may contain images that are associated with dead/running containers, so they cannot be deleted. This means that the number of images actually deleted per execution may not be very many.
* GC runs every 5 minutes, so it can take a really long time before space is freed up when only a handful of images are deleted each time it runs.
 
## expectations
* GC should exhaustively query API for all images, every cycle (it appears to query for about 20).
* GC should iterate over images, attempting to delete them, until low-watermark is reached.
 
## related
* https://github.com/kubernetes/kubernetes/issues/27169

## versions
```
Client Version: version.Info{Major:"1", Minor:"3", GitVersion:"v1.3.0", GitCommit:"283137936a498aed572ee22af6774b6fb6e9fd94", GitTreeState:"clean", BuildDate:"2016-07-01T19:26:38Z", GoVersion:"go1.6.2", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"3", GitVersion:"v1.3.0", GitCommit:"283137936a498aed572ee22af6774b6fb6e9fd94", GitTreeState:"clean", BuildDate:"2016-07-01T19:19:19Z", GoVersion:"go1.6.2", Compiler:"gc", Platform:"linux/amd64"}
```

## reproduction

1. Start a 10-15 containers using different images to simulate "undeleteable" images
2. Download a bunch of images, but never instantiate them: `i=10089; while true; do i=$((i+1)); docker pull private-repo/private-image:master-$i; done`
2. Wait for disk to fill up (or use `dd` to speed up the process!)
3. Observe evictions

Relevant logs:
```
Jul 19 01:15:15 ip-10-21-100-11.ec2.internal bash[4021]: I0719 01:15:15.304394    4021 image_manager.go:234] [ImageManager]: Disk usage on "/dev/xvdf" (/vol/docker) is at 92% which is over the high threshold (90%). Trying to free 11903481446 bytes
Jul 19 01:15:15 ip-10-21-100-11.ec2.internal bash[4021]: I0719 01:15:15.400741    4021 image_manager.go:295] [ImageManager]: Removing image "77eb8d6be627674bcada2e576861cdb38d9c0ee2c1483365de3416b1ce1bf767" to free 817282446 bytes
Jul 19 01:15:15 ip-10-21-100-11.ec2.internal bash[4021]: I0719 01:15:15.405438    4021 image_manager.go:295] [ImageManager]: Removing image "5b9cd8b4cfe7a42deffefbc7510e85c710eb50a9edb1bf1ff55d13e739e72c95" to free 255806935 bytes
Jul 19 01:15:15 ip-10-21-100-11.ec2.internal bash[4021]: I0719 01:15:15.405984    4021 image_manager.go:295] [ImageManager]: Removing image "b07fe51dc2a86888cebbbfbd1e0f5ca97d78d53ed4ac377f2bdcab02cbb64b2d" to free 23418326 bytes
Jul 19 01:15:15 ip-10-21-100-11.ec2.internal bash[4021]: I0719 01:15:15.406735    4021 image_manager.go:295] [ImageManager]: Removing image "34ad95c5989db379437b655d85cdfc2aa1b3c932935dc3085941c3482c0bbdfa" to free 416244985 bytes
Jul 19 01:15:15 ip-10-21-100-11.ec2.internal bash[4021]: I0719 01:15:15.407270    4021 image_manager.go:295] [ImageManager]: Removing image "2fd4336636c3ca6740cb477a199ce0ab91321956c9a5ab2155f476d628855506" to free 211010533 bytes
Jul 19 01:15:15 ip-10-21-100-11.ec2.internal bash[4021]: I0719 01:15:15.407829    4021 image_manager.go:295] [ImageManager]: Removing image "b059cc8d0a14e452742628b1859d37af529fea43dd462307c4990833d23fc014" to free 18864800 bytes
Jul 19 01:15:15 ip-10-21-100-11.ec2.internal bash[4021]: I0719 01:15:15.408335    4021 image_manager.go:295] [ImageManager]: Removing image "396b5650032c4e8268a93d91db9a97a70b0001284cb48106cfee5bdfefce3fcb" to free 738024199 bytes
Jul 19 01:15:17 ip-10-21-100-11.ec2.internal bash[4021]: I0719 01:15:17.731291    4021 image_manager.go:295] [ImageManager]: Removing image "880876037199ec51faa913ff0fe170bd3588d051e1de8ed340175998806fc626" to free 655183184 bytes
Jul 19 01:15:19 ip-10-21-100-11.ec2.internal bash[4021]: I0719 01:15:19.056107    4021 image_manager.go:295] [ImageManager]: Removing image "70d207217affe0958f88207140cb6e17b648f493bee2894d86203aebcc365877" to free 738041863 bytes
Jul 19 01:15:21 ip-10-21-100-11.ec2.internal bash[4021]: I0719 01:15:21.365380    4021 image_manager.go:295] [ImageManager]: Removing image "ccfc30e62d621a4787bf1f1e4adcb1bc8b1d6c215fac1ab927b79ddc0beaf74a" to free 738037445 bytes
Jul 19 01:15:23 ip-10-21-100-11.ec2.internal bash[4021]: I0719 01:15:23.562847    4021 image_manager.go:295] [ImageManager]: Removing image "f2c4d246b910ab0667d8c3e7f92ef51d274ab99c58c64598cb490bd213230201" to free 738038872 bytes
Jul 19 01:20:25 ip-10-21-100-11.ec2.internal bash[4021]: I0719 01:20:25.687659    4021 image_manager.go:234] [ImageManager]: Disk usage on "/dev/xvdf" (/vol/docker) is at 92% which is over the high threshold (90%). Trying to free 11472557670 bytes
Jul 19 01:20:25 ip-10-21-100-11.ec2.internal bash[4021]: I0719 01:20:25.785083    4021 image_manager.go:295] [ImageManager]: Removing image "b07fe51dc2a86888cebbbfbd1e0f5ca97d78d53ed4ac377f2bdcab02cbb64b2d" to free 23418326 bytes
Jul 19 01:20:25 ip-10-21-100-11.ec2.internal bash[4021]: I0719 01:20:25.787988    4021 image_manager.go:295] [ImageManager]: Removing image "34ad95c5989db379437b655d85cdfc2aa1b3c932935dc3085941c3482c0bbdfa" to free 416244985 bytes
Jul 19 01:20:25 ip-10-21-100-11.ec2.internal bash[4021]: I0719 01:20:25.788595    4021 image_manager.go:295] [ImageManager]: Removing image "2fd4336636c3ca6740cb477a199ce0ab91321956c9a5ab2155f476d628855506" to free 211010533 bytes
Jul 19 01:20:25 ip-10-21-100-11.ec2.internal bash[4021]: I0719 01:20:25.789330    4021 image_manager.go:295] [ImageManager]: Removing image "77eb8d6be627674bcada2e576861cdb38d9c0ee2c1483365de3416b1ce1bf767" to free 817282446 bytes
Jul 19 01:20:25 ip-10-21-100-11.ec2.internal bash[4021]: I0719 01:20:25.789935    4021 image_manager.go:295] [ImageManager]: Removing image "5b9cd8b4cfe7a42deffefbc7510e85c710eb50a9edb1bf1ff55d13e739e72c95" to free 255806935 bytes
Jul 19 01:20:25 ip-10-21-100-11.ec2.internal bash[4021]: I0719 01:20:25.790646    4021 image_manager.go:295] [ImageManager]: Removing image "b059cc8d0a14e452742628b1859d37af529fea43dd462307c4990833d23fc014" to free 18864800 bytes
Jul 19 01:20:25 ip-10-21-100-11.ec2.internal bash[4021]: I0719 01:20:25.791200    4021 image_manager.go:295] [ImageManager]: Removing image "3c5a14b717b7e0f2b65b9ebb0620540f88d0f88a0e0de78571c911302b6dfcaa" to free 765846156 bytes
Jul 19 01:20:26 ip-10-21-100-11.ec2.internal bash[4021]: I0719 01:20:26.653396    4021 image_manager.go:295] [ImageManager]: Removing image "2a0ad5140baaba6ec692d96793f88b511ff172eca6a1d3798e73bc7c21df6d79" to free 765852838 bytes
Jul 19 01:20:27 ip-10-21-100-11.ec2.internal bash[4021]: I0719 01:20:27.956725    4021 image_manager.go:295] [ImageManager]: Removing image "200b9adf5227da6c481ce59f10b82d0d83585fdad35e8ca4b6c6e6b7fa43309f" to free 655183184 bytes
Jul 19 01:20:29 ip-10-21-100-11.ec2.internal bash[4021]: I0719 01:20:29.594989    4021 image_manager.go:295] [ImageManager]: Removing image "331bbc1fe0a6e5ac250b281b907a01224b88e1f1c20537fd2d7c178eb39072b5" to free 655183184 bytes
Jul 19 01:20:31 ip-10-21-100-11.ec2.internal bash[4021]: I0719 01:20:31.882156    4021 image_manager.go:295] [ImageManager]: Removing image "4c2d8185ea6cab305a73c6577e656af069f0e1733445ff2daa7618bd3fcf0297" to free 765844450 bytes
Jul 19 01:25:32 ip-10-21-100-11.ec2.internal bash[4021]: I0719 01:25:32.729735    4021 image_manager.go:234] [ImageManager]: Disk usage on "/dev/xvdf" (/vol/docker) is at 94% which is over the high threshold (90%). Trying to free 14006699622 bytes
Jul 19 01:25:32 ip-10-21-100-11.ec2.internal bash[4021]: I0719 01:25:32.837732    4021 image_manager.go:295] [ImageManager]: Removing image "77eb8d6be627674bcada2e576861cdb38d9c0ee2c1483365de3416b1ce1bf767" to free 817282446 bytes
Jul 19 01:25:32 ip-10-21-100-11.ec2.internal bash[4021]: I0719 01:25:32.842331    4021 image_manager.go:295] [ImageManager]: Removing image "34ad95c5989db379437b655d85cdfc2aa1b3c932935dc3085941c3482c0bbdfa" to free 416244985 bytes
Jul 19 01:25:32 ip-10-21-100-11.ec2.internal bash[4021]: I0719 01:25:32.842935    4021 image_manager.go:295] [ImageManager]: Removing image "b059cc8d0a14e452742628b1859d37af529fea43dd462307c4990833d23fc014" to free 18864800 bytes
Jul 19 01:25:32 ip-10-21-100-11.ec2.internal bash[4021]: I0719 01:25:32.843505    4021 image_manager.go:295] [ImageManager]: Removing image "2fd4336636c3ca6740cb477a199ce0ab91321956c9a5ab2155f476d628855506" to free 211010533 bytes
Jul 19 01:25:32 ip-10-21-100-11.ec2.internal bash[4021]: I0719 01:25:32.844061    4021 image_manager.go:295] [ImageManager]: Removing image "5b9cd8b4cfe7a42deffefbc7510e85c710eb50a9edb1bf1ff55d13e739e72c95" to free 255806935 bytes
Jul 19 01:25:32 ip-10-21-100-11.ec2.internal bash[4021]: I0719 01:25:32.844625    4021 image_manager.go:295] [ImageManager]: Removing image "b07fe51dc2a86888cebbbfbd1e0f5ca97d78d53ed4ac377f2bdcab02cbb64b2d" to free 23418326 bytes
Jul 19 01:25:32 ip-10-21-100-11.ec2.internal bash[4021]: I0719 01:25:32.845236    4021 image_manager.go:295] [ImageManager]: Removing image "7d5384f80e5d6d3cad9f6ed99de953f9b0c909e06402ada4026aa31bdeb8ed10" to free 655183184 bytes
Jul 19 01:25:35 ip-10-21-100-11.ec2.internal bash[4021]: I0719 01:25:35.277384    4021 image_manager.go:295] [ImageManager]: Removing image "ac8b2be5a4c1cee99a22eb74fb94013b39ef6a7852d01d131ffbe51c74dc13c5" to free 655183184 bytes
Jul 19 01:25:37 ip-10-21-100-11.ec2.internal bash[4021]: I0719 01:25:37.586057    4021 image_manager.go:295] [ImageManager]: Removing image "11eb93f63d823f496145e9755e4867f0b64f00b089af359a5497c92d3cbacb09" to free 655183184 bytes
Jul 19 01:25:40 ip-10-21-100-11.ec2.internal bash[4021]: I0719 01:25:40.614358    4021 image_manager.go:295] [ImageManager]: Removing image "c915bc31d47c474b62f14f49551d1fcd2c25380355702d8cc5c45dd6411cd275" to free 655183184 bytes
Jul 19 01:25:43 ip-10-21-100-11.ec2.internal bash[4021]: I0719 01:25:43.456947    4021 image_manager.go:295] [ImageManager]: Removing image "2229e668bcd357473fa2fc4df2724798b8c912b077a8d6052af6642e4b64db98" to free 642975123 bytes
```

**Here's what I think is happening:**
Every GC execution (approx every 5 minutes), less than ~20 images are pulled from the docker API as candidates for deletion. From the output, we can see that every time a subset of the image id candidates are the same and a subset are new. Those repeated images are the ones that cannot be deleted for one reason or another (e.g. `conflict: unable to delete b07fe51dc2a8 (cannot be forced) - image is being used by running container 7767cc4392a8` - confirmed this is an example). Then the remaining images are probably new candidates, some of which will probably removed. I think GC might be working, albeit *very*, *very* slowly. Also, it would appear to me that if I had more than 20 containers running on this server, that there would be no images successfully deleted because it keeps including them in every attempt.  

I have also confirmed that there are more images available for deletion by doing `docker rmi` on all images to free up space. 

To prove there are images that can be deleted, I ran:
```
docker images -a | awk '{print $3}' | xargs -n 1 docker rmi
```

Amongst the output, I saw the following: (and many more)
```
Deleted: 72e59eceed148802db5d11a0222fed9d1618a4af9236af6e07ee199917f43457
Deleted: 118e38b974b6d971d708ac47e5a578d11cc89c07ae4e6e286a454c30bda7207f
Deleted: 6be1e24bb03a7fdfffac664281b920caba8ffc0fe2d7180d8a9e349ca14f242a
Deleted: acd0b6901b48aed75c8a30fedf3547b0580c199fc85a8211bec4451e7fde374a
Deleted: 89a61bc317d5ba4fe755ea303fd94f747cfeef4e5fc935613f414c0c424b78dc
Deleted: 80f404a5dc273dfe981d5e640e7d963ef8b6ece6445fc42ca37e1a3b12b58c44
Deleted: 6b6b4d6e9810e34586acef12bbdf66225d6cfef4a74147658b434b194fd723bf
Deleted: 8b0c71763ca3184c5ccc998ab787af754ab92ce86c616ea2b0ca3d95503db712
Deleted: 9bb223f8fea8f78277621e49fdb482c6d8300f09b97d642f202372085dc95bdf
Deleted: 91f4237da9015be843c1f03b97af818423dacf78c39b65f452b53b1db249c42e
Deleted: e3adbb2fe9420ff423da30e160a3fabbac4f5e32bc54e02d08ca7ba958273195
Deleted: 83ecb428b82ae6f5114fbe0d1725f89e6ed445bdee6273d6201f28540205f1e7
Deleted: 4f8ffe1a0fdb804b085e03ab695ee4b72dd5cb370604b1067da5bc30e60a0b04
Deleted: 552459980809ab2a64c0607e73a25a0a0e868ed0983f0f9d832e9b1e0f9be176
```

**Comment by yujuhong** [comment]
@osterman It'd be nice if you can provide the kubelet/docker version you use. Thanks!

**Comment by osterman** [comment]
@yujuhong thanks - forgot to add that. updated description with version.

**Comment by yujuhong** [comment]
@osterman, @ronnielai has fixed several image GC issues (see #29316) in 1.4 and 1.3. If this is still happening in your cluster, please file a new issue. Thanks!
